### DISCUSSION TITLE
Time asynchronism error between IMU and GPS

### DISCUSSION THREAD

**Discussion by Boooooots** [rodralez/NaveGo#90] [Q&A] [answered]
Hi, 
    I found that the time asynchronism error between IMU and GPS was not dealt with in the project. I want to know how much impact the time asynchronism error has on the performance and accuracy of the whole integrated navigation in actual operation.  thank you.

**Comment by rodralez** [comment, marked as answer]
IMU and GNSS sensors usually work at different sampling frequencies and the data delivery of both measurements is not synchronized either. And, in fact, this is not a problem. Each sensor has its own timestamp vector (`imu.t` and `gnss.t`), Navego compares both vectors at each imu.t timestamp. If a new GNSS data is available inside a time window ε (epsilon), where the inertial current timestamp is centered in the middle, the Kalman filter is executed. If not, the inertial navigation system keeps executing. See `ins_gnss.m` [line 315](https://github.com/rodralez/NaveGo/blob/0694e86426f09f09fb355424424d2eecc6c379a0/ins-gnss/ins_gnss.m#L315).

I hope these comments will be helpful.

Kind regards.

**Reply by Boooooots** [reply]
Thank you for your reply, your matlab code is equivalent to implement post-processing fusion, can by comparing the timestamp, but if it is real-time processing of IMU and GPS data on the embedded devices, because of the GPS and IMU data lag, the existence of IMU and GPS from sampling to the moment to algorithm to perform the moment there are differences between the time lag.  How to deal with the error caused by the time lag difference here?  
Thank you again.

**Reply by rodralez** [reply]
In my opinion, the time lag between IMU and GNSS at the Kalman filter execution step, has little to no influence in the overall navigation system performance, at least when using tactical- or consumer-grade sensors. You can use the maximum IMU sampling frequency to reduce this lag time. Nevertheless, it is almost impossible for IMU and GNSS data to arrive at the same time. 

NaveGo's approach is the same as a real-time navigation system. When new GNSS data is available, the Kalman filter is run using the last INS updated variables. Moreover, the lag time between IMU and GNSS cannot be longer than the time window ε. So, you can set your tolerance to IMU-GNSS time lag.

**Reply by Boooooots** [reply]
I see, thank you.
